### PR TITLE
Enable reconfiguration of the DeviceNode GW

### PR DESCRIPTION
**Pull request by kursatkobya**
Add isConfigured variable to devicenodeparser.h
Add component tests for reconfiguring DeviceNodeGateway

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

### CODE REVIEW

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.h [hunk 1]
```diff
@@ -70,6 +70,13 @@ class DeviceNodeGateway :
      */
     virtual bool teardownGateway() override;
 
+    /**
+     * @brief Questions whether the device is configured or not
+     *
+     * return true or false regarding configuration status
+     */
+    virtual bool isDeviceConfigured(const std::string deviceName);
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.h]: Shouldn't this be private instead of public?

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp [hunk 2]
```diff
@@ -51,38 +51,41 @@ bool DeviceNodeGateway::activateGateway()
...
-                log_error() << "Could not 'chmod " << dev.mode
-                            << "' the mounted device " << dev.name;
+            // If the parent directory does not already exist, we create it.
+            if (!isDirectory(deviceParent) && !createDirectory(deviceParent)) {
+                log_error() << "Could not create parent directory for device " << deviceParent;
                 return false;
             }
+
+            // Mount device in container
+            getContainer()->mountDevice(dev.name);
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp]: Shouldn't there be some error checking here?

File: libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp [hunk 3]
```diff
@@ -97,3 +97,60 @@ TEST_F(DeviceNodeGatewayTest, TestOverwriteDeviceFails) {
...
+            "/dev/tty0",
+            true},
+        testSetup{
+            "[{\"name\": \"/dev/tty1\", \"mode\":  622}]",
+            "[{\"name\": \"/dev/tty1\", \"mode\":  622}]",
+            "/dev/tty1",
+            true}
+));
+
+TEST_P(DynamicConfigurationTests, DynamicDeviceNodeConfiguration) {
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp]: Needs a comment what the test is about etc.

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp [hunk 4]
```diff
@@ -97,3 +97,60 @@ TEST_F(DeviceNodeGatewayTest, TestOverwriteDeviceFails) {
...
+
+/*
+ * This data is fed to the DynamicConfiguration tests
+ */
+INSTANTIATE_TEST_CASE_P(DeviceModeConfigurations, DynamicConfigurationTests, ::testing::Values(
+        testSetup{
+            "[{\"name\": \"/dev/tty0\", \"mode\":  622}]",
+            "[{\"name\": \"/dev/tty0\", \"mode\":  755}]",
+            "/dev/tty0",
+            false},
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp]: For readability, could you add a space before the opening braces and break the line for the closing braces and fix the indentation so the config strings are indented one level and the braces matches on the outer level?

Like:

testSetup {
    "string",
    "string"
},
testSetup {

etc.

File: libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp [hunk 5]
```diff
@@ -97,3 +97,60 @@ TEST_F(DeviceNodeGatewayTest, TestOverwriteDeviceFails) {
     checkMode.wait();
     ASSERT_TRUE(checkMode.isSuccess());
 }
+
+struct testSetup
+{
+    const std::string firstConfiguration;
+    const std::string secondConfiguration;
```
Comment (tobsan) [inline libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp:104]: It would be nice to be able to use more than two configuration stages in the tests, since the implementation is not limited to two afaict.

**Review by tobsan — CHANGES_REQUESTED** [review]
Looks good, just had some comments. I like the general idea to mark a device as up-to-date or not.

File: libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp (hunk 5 above)
Comment (breakreturn) [reply libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp:104]: This is a good point, the quick-launch feature might not require more than two sequences at the moment, but our support for this in general is not so far limited to two and I don't think this should be either.

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.h (hunk 1 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.h]: This function is needed by gtest to examine the idea of "only update device when it is needed".

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp (hunk 2 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp]: this part is not new code, but sure.

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp [hunk 6]
```diff
@@ -51,38 +51,45 @@ bool DeviceNodeGateway::activateGateway()
...
 
     for (auto &dev : devlist) {
-        log_info() << "Mapping device " << dev.name;
+        if (!dev.isConfigured) {
+            log_info() << "Mapping device " << dev.name;
 
-        std::string devicePathInContainerOnHost = buildPath(getContainer()->rootFS(), dev.name);
-        std::string deviceParent = parentPath(devicePathInContainerOnHost);
+            std::string devicePathInContainerOnHost = buildPath(getContainer()->rootFS(), dev.name);
+            std::string deviceParent = parentPath(devicePathInContainerOnHost);
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp]: I just realized that this whole check for `deviceParent` etc only works because we are lucky. The /dev/ directory inside the container is mounted as a tmpfs, and as such, we can't see it from the host side.

@kursatkobya can you update the test to try to mount `/dev/input/event0` into the container? I suspect this should fail (I should've taken care of this when I handled the general bind-mounting issue) because `/dev/input` does not exist inside the container, and the `createDirectory` calls will only create it on the host (not inside the tmpfs where `/dev` in the container resides).

If that is the case, we should update to try to create any needed parent directories inside the container (with a functionjob, I guess).

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp [hunk 7]
```diff
@@ -51,38 +51,29 @@ bool DeviceNodeGateway::activateGateway()
     }
 
     for (auto &dev : devlist) {
-        log_info() << "Mapping device " << dev.name;
-
-        std::string devicePathInContainerOnHost = buildPath(getContainer()->rootFS(), dev.name);
-        std::string deviceParent = parentPath(devicePathInContainerOnHost);
-
-        // Already existing files can't be converted into directories
-        if (existsInFileSystem(deviceParent) && !isDirectory(deviceParent)) {
```
Comment (frznlogic) [inline libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp:60]: Are these checks performed somewhere else or why are we removing them?

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.h (hunk 1 above)
Comment (frznlogic) [reply libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.h]: Could it be made protected instead in that case?

**Review by frznlogic — APPROVED** [review]

File: libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp (hunk 7 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp:60]: Those lines are removed, because we are no longer support creating and mounting new devices.

Comment (tobsan) [reply libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp:60]: or rather, when we add an existing device, LXC handles creation of any parent directories.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+Device::Device(const Device &dev) :
+        name(dev.name),
+        mode(dev.mode),
+        isConfigured(dev.isConfigured)
+{
+}
+
+bool Device::parse(const json_t *element)
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp]: I would prefer if this function was a static factory function, so that it creates a Device object on successful parse.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.h [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+
+namespace softwarecontainer {
+
+class Device
+{
+    LOG_DECLARE_CLASS_CONTEXT("DN", "Device Node");
+
+public:
+    Device(std::string _name = "", int _mode = -1, bool _condigured = false);
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenode.h]: I don't think we need defaults for these, do we?

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.h [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * @brief Activates this device by running mknod and chmod commands which are run in the
+     * container.
+     *
+     * @return true upon success, false otherwise
+     */
+    bool activate(std::shared_ptr<ContainerAbstractInterface> container);
+
+    std::string name;
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenode.h]: This should be private with getter/setter if needed.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.h [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    /*
+     * @brief Holds device mode which consist 3 integer number between 0 and 7 which represents
+     * permissions for owner, groups and all users
+     * */
+    int mode;
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenode.h]: This should be private with getter/setter if needed.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.h [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+    /*
+    * @brief Holds the flag that whether the device is configured or not
+    */
+    bool isConfigured;
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenode.h]: This should be private with getter/setter if needed.

File: libsoftwarecontainer/src/gateway/devicenode/devicenodelogic.cpp [hunk 13]
```diff
@@ -18,33 +18,40 @@
...
+std::vector<std::shared_ptr<Device>>::iterator
 DeviceNodeLogic::findDeviceByName(const std::string name)
 {
-    std::vector<DeviceNodeParser::Device>::iterator item =
+    std::vector<std::shared_ptr<Device>>::iterator item =
             std::find_if(m_devList.begin(), m_devList.end(),
-            [&] (DeviceNodeParser::Device const &d) { return d.name == name; });
+            [&] (std::shared_ptr<Device> const &d) { return d->name == name; });
 
     return item;
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenodelogic.cpp]: I think it's a bit weird to return an iterator to a shared_ptr. Can't you return the shared_ptr directly instead?

File: libsoftwarecontainer/src/gateway/devicenode/devicenodelogic.cpp [hunk 14]
```diff
@@ -18,33 +18,40 @@
...
+        std::shared_ptr<Device> newDevice(new Device(dev));
+        newDevice->isConfigured = false;
+        m_devList.push_back(std::move(newDevice));
     } else {
-        item->mode = calculateDeviceMode(item->mode, dev.mode);
+        auto mode = calculateDeviceMode(((*item))->mode, dev.mode);
+        if (mode != (*item)->mode) {
+            (*item)->isConfigured = false;
+            (*item)->mode = mode;
+        }
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenodelogic.cpp]: If you return the shared_ptr directly, from findDeviceByName, this code would look way better.

File: libsoftwarecontainer/unit-test/devicenode_unittest.cpp [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+    ASSERT_TRUE(dev.activate(container));
+    ASSERT_TRUE(dev.isConfigured);
+}
+
+
+
+struct testSetup
+{
+    Device firstConfiguration;
+    Device secondConfiguration;
```
Comment (tobsan) [inline libsoftwarecontainer/unit-test/devicenode_unittest.cpp:155]: As I said earlier, it would be nice to be able to test more than two configuration runs. Perhaps make it a vector instead?

**Review by tobsan — CHANGES_REQUESTED** [review]
I really like the re-design of this gateway. However, it looks super weird to hold shared_ptr of Device in DeviceNodeLogic, when the only pointer operations are done inside the logic class itself. That feels completely backwards.

I would propose that the Device class itself can do the whitelisting operations now done in DeviceNodeLogic, or that the lookup of devices in the logic class is changed if possible.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.h (hunk 9 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/gateway/devicenode/devicenode.h]: except iscondigured it would be better to have defaults imo.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp (hunk 8 above)
Comment (kursatkobya) [reply libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp]: we can overload constructor but i think this way is easier to read.

File: libsoftwarecontainer/unit-test/devicenode_unittest.cpp (hunk 15 above)
Comment (kursatkobya) [reply libsoftwarecontainer/unit-test/devicenode_unittest.cpp:155]: We can create another test to examine this. But what would have been the perks of having it ? Since this feature is provided by inherited Gateway class, it is actually on/off option. Either a gateway can be activated once or no limitation.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+#include "devicenode.h"
+#include "functionjob.h"
+
+namespace softwarecontainer {
+
+Device::Device(std::string _name, int _mode) :
+        name(_name),
+        mode(_mode),
+        isConfigured(false)
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp]: Perhaps nitpicking, but you should use m_name, m_mode and m_isConfigured for the members. Then you won't need _name etc.

File: libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+
+int Device::getMode()
+{
+    return mode;
+}
+
+bool Device::getIsconfigured ()
+{
+    return isConfigured;
+}
```
Comment (tobsan) [inline libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp]: I would prefer if the getters had the name of the variable, so mode(), name() etc.

File: libsoftwarecontainer/unit-test/devicenode_unittest.cpp (hunk 15 above)
Comment (tobsan) [reply libsoftwarecontainer/unit-test/devicenode_unittest.cpp:155]: I just mean, instead of testing one or two activations, we could test one or more activations.

File: libsoftwarecontainer/unit-test/devicenode_unittest.cpp [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+
+/*
+ * Verify whitelisting in the case when mode is set to more permissive
+ */
+TEST_P(DeviceModeTests, UpdateDeviceMode) {
+    auto testparams = GetParam();
+    dev.setMode(testparams.itemMode);
+    dev.calculateDeviceMode(testparams.newMode);
+    ASSERT_EQ(testparams.expectedMode, dev.getMode());
+}
```
Comment (tobsan) [inline libsoftwarecontainer/unit-test/devicenode_unittest.cpp:249]: These tests perform the same testing as the component test that's added in this PR does, right? Then I would vote for removing the component test. We should add a service test, but that should not be part of this PR imo.

**Review by tobsan — CHANGES_REQUESTED** [review]
Some new comments.

File: libsoftwarecontainer/unit-test/devicenode_unittest.cpp (hunk 18 above)
Comment (kursatkobya) [reply libsoftwarecontainer/unit-test/devicenode_unittest.cpp:249]: The component tests additionally performs mount operations. So i think we should remove them after adding some related service tests.

**Review by tobsan — APPROVED** [review]
